contract-sdk/types/src/crypto.rs

```diff
@@ -0,0 +1,77 @@
+use k256::{
+    ecdsa::{recoverable, Signature},
+    EncodedPoint,
+};
+use thiserror::Error;
+
+/// ECDSA recover error.
+#[derive(Error, Debug, cbor::Encode, cbor::Decode)]
+pub enum Error {
+    #[error("malformed input")]
+    MalformedInput,
+    #[error("malformed signature")]
+    MalformedSignature,
+    #[error("signature verification failed")]
+    VerificationFailed,
+}
+
+/// ECDSA public key recovery method.
+pub fn ecdsa_recover(input: &[u8]) -> Result<Vec<u8>, Error> {
+    // [32] hash, [32] r, [32] s, [1] v.
+    if input.len() != 97 {
+        return Err(Error::MalformedInput);
+    }
+
+    let mut msg = [0u8; 32];
+    let mut r = [0u8; 32];
+    let mut s = [0u8; 32];
+    let v = input[96];
+
+    msg[0..32].copy_from_slice(&input[0..32]);
+    r[0..].copy_from_slice(&input[32..64]);
+    s[0..].copy_from_slice(&input[64..96]);
+
+    let signature = Signature::from_scalars(r, s).map_err(|_| Error::MalformedSignature)?;
+    let signature = recoverable::Signature::new(
+        &signature,
+        recoverable::Id::new(v).map_err(|_| Error::MalformedSignature)?,
+    )
+    .map_err(|_| Error::MalformedSignature)?;
+
+    if signature.s().is_high().into() {
+        return Err(Error::MalformedSignature);
+    }
+
+    match signature.recover_verify_key_from_digest_bytes(&msg.into()) {
+        Ok(recovered_key) => Ok(EncodedPoint::from(&recovered_key)
+            .decompress()
+            .unwrap()
+            .as_bytes()
+            .to_vec()),
```

Comment:
I believe the return type could be fixed size, right? This way we could avoid the need for allocations from the host as the caller could just preallocate an empty array and pass its address. I guess in this case we would also need something to signal an error.

Reply:
Yeah it could be done that way. So we could return a zero array in case something failed (that's what i currently do anyway in the "link function"), with a todo to propagate the error :-) - not sure if having the concrete error in the contract is all that useful anyway.

Reply:
Only thing the error would really tell you if the signature is invalid, or if the whole input is malformed. I see that returning the zero array in case the verification failed is also what we do in the evm precompile.

Reply:
Yeah this makes sense.